packages/permissions/src/PermissionsAPI.ts

```diff
@@ -0,0 +1,213 @@
+    PermissionsAPI.#events[_handler.eventId] = {}
+
+    return _handler
+  }
+
+  static #getAsyncHandler(
+    permissionOption: PermissionOption
+  ): IAsyncPermissionHandler {
+    const _handler: IAsyncPermissionHandler =
+      async function (): Promise<AsyncPermissionResponse> {
+        const _permission = await PermissionsAPI.getPermission(permissionOption)
+        const _events = PermissionsAPI.#events[_handler.eventId]
+
+        return new Promise<AsyncPermissionResponse>((resolve, reject) => {
+          if ('status' in _permission) {
+            if (_events.onPermissionChange) {
+              const onChange = _events.onPermissionChange
+              _permission.status.onchange = function () {
+                onChange({
+                  message: this.name,
+                  status: this,
+                  state: this.state
+                })
+              }
+            }
+
+            if (_permission.state === 'denied') {
+              if (_events.onPermissionDenied)
+                _events.onPermissionDenied(
+                  _permission as PermissionResponse<'denied'>
+                )
+
+              return resolve({
+                granted: null,
+                denied: _permission as PermissionResponse<'denied'>
+              })
+            } else {
+              if (_events.onPermissionGranted)
+                _events.onPermissionGranted(
+                  _permission as PermissionResponse<'granted'>
+                )
+
+              return resolve({
+                granted: _permission as PermissionResponse<'granted'>,
+                denied: null
+              })
+            }
+          } else {
+            if (_events.onPermissionError)
+              _events.onPermissionError(
+                _permission as PermissionResponse<'error'>
+              )
+
+            return reject(_permission as PermissionResponse<'error'>)
+          }
+        })
+      }
+    _handler.eventId = crypto.randomUUID()
+    PermissionsAPI.#events[_handler.eventId] = {}
+
+    return _handler
+  }
+
+  static async getPermission(
+    option: PermissionOption
+  ): Promise<PermissionResponse | PermissionResponse<'error'>> {
+    try {
+      const _status = await PermissionsAPI.#permissions.query(option as any)
+      const permissionStatus = {
+        state: _status.state,
+        status: _status,
+        message: _status.name
+      }
+      permissionStatus.status.addEventListener('change', function () {
+        PermissionsAPI.#permissionStates[option.name].state = this.state
+      })
+      PermissionsAPI.#permissionStates[option.name] = permissionStatus
+
+      return PermissionsAPI.#permissionStates[option.name]
+    } catch (error: any) {
+      if (error.name === 'TypeError') {
+        return {
+          state: 'unsupported',
+          message: error.message
+        }
+      }
+
+      return {
+        state: 'invalid',
+        message: error.message
+      }
```

Comment:
The `getPermission` method lacks proper cleanup of the event listener when the permission state changes, which can lead to memory leaks.

```diff
-      permissionStatus.status.addEventListener('change', function () {
+      const changeListener = function () {
         PermissionsAPI.#permissionStates[option.name].state = this.state;
-      })
+      };
+      permissionStatus.status.addEventListener('change', changeListener);
+      permissionStatus.cleanup = () => permissionStatus.status.removeEventListener('change', changeListener);
```